ui/src/views/Documentation/index.jsx

```diff
@@ -106,63 +182,220 @@ export default class Documentation extends Component {
   }
 
+  buildSchemaId(schemaId) {
+    if (schemaId.startsWith('/')) {
+      if (
+        process.env.TASKCLUSTER_ROOT_URL &&
+        process.env.TASKCLUSTER_ROOT_URL !== 'https://taskcluster.net'
+      ) {
+        return process.env.TASKCLUSTER_ROOT_URL + schemaId;
+      }
+
+      return `https://schemas.taskcluster.net/${schemaId.replace(
+        /^\/schemas\//,
+        ''
+      )}`;
+    }
+
+    return schemaId;
+  }
+
+  sanitizeSchema(schema) {
+    if (schema.$id) {
+      return {
+        ...schema,
+        $id: this.buildSchemaId(schema.$id),
+      };
+    }
+
+    if (schema.id) {
+      return {
+        ...schema,
+        id: this.buildSchemaId(schema.id),
+      };
+    }
+
+    return schema;
+  }
+
+  // Returns a mapping between the HTML element and the desired component
+  components() {
+    return {
+      a: this.anchorFactory,
+      h1: this.headingFactory('h1'),
+      h2: this.headingFactory('h2'),
+      h3: this.headingFactory('h3'),
+      h4: this.headingFactory('h4'),
+      h5: this.headingFactory('h5'),
+      h6: this.headingFactory('h6'),
+      img: this.imageFactory,
+    };
+  }
+
   async load() {
     try {
-      const { default: Document } = await this.handleImport(
-        this.props.match.params.path
+      const { params } = this.props.match;
+      const pathname = params.path || 'index';
+      const isJsonFile =
+        pathname.startsWith('reference') &&
+        (pathname.endsWith('api') || pathname.endsWith('events'));
+      const { loader, path } = readDocFile(
+        `${pathname}.${isJsonFile ? 'json' : 'md'}`
       );
+      const { default: Page } = await loader;
+
+      absolutePath = path;
+
+      // use `isMDXComponent` once https://github.com/mdx-js/mdx/pull/369 is merged
+      if (!Page.prototype) {
+        const entries = await this.getReferenceEntries(Page.entries);
+
+        return this.setState({
+          Page: null,
+          pageInfo: null,
+          error: null,
+          referenceJson: Object.assign({}, Page, { entries }),
+        });
+      }
+
       const pageInfo = this.getPageInfo();
 
-      this.setState({ Document, pageInfo, error: null });
+      this.setState({ Page, pageInfo, error: null, referenceJson: null });
     } catch (error) {
       this.setState({ error });
     }
   }
 
-  async handleImport(url) {
-    const doc = url ? url.replace(/\/$/, '') : 'index';
+  getDocument(Page) {
+    if (!Page) {
+      return null;
+    }
 
-    return import(/* webpackChunkName: 'Documentation.page' */ `../../docs/${doc}.md`).catch(
-      () =>
-        import(/* webpackChunkName: 'Documentation.page' */ `../../docs/${doc}/index.md`)
+    const page = renderToString(
+      <BrowserRouter>
+        <Page components={this.components()} />
+      </BrowserRouter>
     );
+
+    return page
+      .split(DOCS_SCHEMA_REGEX)
+      .map(html =>
+        !isUrl(html) && html.startsWith('<') ? (
```

Comment:
Maybe I'm misunderstanding this, but I think this is splitting on a regexp that matches a group and thus returns an array of `[ unmatched, matchGroup, unmatched, matchGroup, .. ]`, then "guessing" whether each element is unmatched data (because it doesn't look like a URL but does look like HTML in that it begins with `<`) or a URL.  In fact, it adds a dependency to help with that guessing.  But the positions of the unmatched data and the matchGroup are fixed within the array, so you could use `.map((val, i) => (i & 1 == 0) ? val : <SchemaTable ..>)`.

Couldn't this be handled with a `components` entry, anyway?  We are not committed to using `data-render-schema` -- we could easily change all of those to e.g., `<jsonSchema path="/schemas/notify/v1/irc-request.json" />`.  I suspect that would save a lot of difficulty breaking HTML into pieces and re-assembling it with `Fragment` and `dangerouslySetInnerHtml`!

Reply:
Crazy idea: could we handle *everything* as a markdown document, and generate some of them from other data structures during generation?  So for example `api.md` would just be `<reference path="/references/auth/v1/api.json" />`, and there a a `components` entry for `reference` that knows how to render that reference based on data in `generated/references.json`.

Then the entire `/docs` chunk is just displaying a tree of Page objects, each one like the next.

Reply:
@djmitche 

> Crazy idea: could we handle _everything_ as a markdown document

Yes, we definitely can. What should I change so that in addition to api.json and events.json, services also get `api.md` and `events.md`? I'm thinking of having the markdown be:

```
---
title: ...
order: 1
inline: true
---
import Reference from 'path/to/the/Reference/component'
import api from './api.json'

<Reference json={api} />
```

Reply:
I like that!  I think you'd just need to add some support to `yarn generate` to generate such files. 

In their definition, api and exchange (and soon log) references are "sort of" all the same thing.  They are all listed in an array in `manifest.json` and differentiated by getting the schema identified by their $schema property (for example, https://dustin.taskcluster-dev.net/references/queue/v1/exchanges.json points to https://dustin.taskcluster-dev.net/schemas/common/exchanges-reference-v0.json#) and looking at its metadata.name and metadata.version fields.  Perhaps `Reference` can do that.  Or, perhaps `yarn generate` could do that lookup and use distinct components for different `metadata.name`'s (and maybe even helpfully include the version):

```
import ApiReference from '..';

<ApiReference json={api} version={0} />
```

Given all this, it sounds like we could safely generate *everything* under `ui/`, leaving only `references.json` in `generated/` (and I can make that go away if necessary, too).